README.md


-------------------------
### Run Google Colab

https://colab.research.google.com/drive/1OShIMVcFZ_khsUIBOIV1lzrqAGo1gfm_?usp=sharing

-------------------------

Distributed binaries may or may not have one or more of the following features 
enabled.


*********************
* REMOTE MINER SERVER
*********************
Place a text file named banned.txt with IP addresses of banned clients in the 
bitcoin working directory.  One IP address should be entered per line, so each
IP address is separated by a newline.

Remote miner server arguments

-remoteserver
	Turns on the remote server.
	
-remotebindaddr=x.x.x.x
	Bind server to specific adapter.  The default is 127.0.0.1.  Note that this 
	will only accept connections from the local computer.
	
-remotebindport=xxxxx
	Bind server to specific port.  The default is 8335.
	
-remotepassword=xxxxx
	Set a password to access the server.  The default is a blank password.
	
-distributiontype=connected|contributed
	Sets method used to distribute bitcoins.  "connected" will distribute coins
	only to those clients that were connected when the block being solved was
	created.  The distribution is based on each connected clients calculated hash 
	rate against the total hash rate at the time a new block is created. 
	"contributed" will accrue all hashes sent to the server for a given address 
	since the last generated block.  A client may freely disconnect and reconnect 
	and will continue accumulating hashes to whatever address the client specified.
	The distribution of coins with this method is based on the hashes accrued by 
	each address against the total hashes accured by everyone.  The server will
	save the values when it shuts down and load them back up on startup.
	
-resethashescontributed
	Resets the count of hashes contributed from each address.


*********************
* REMOTE MINER CPU CLIENT
*********************
Remote miner client arguments

-server=x.x.x.x
	The address of the server to connect to.  The default is 127.0.0.1.
	
-port=xxxxx
	The port of the server.  The default is 8335.
	
-password=xxxxx
	The password to use when connecting to the server.  The default is a blank 
	password.
	
-address=xxxxxxx
	The bitcoin address you want generated coins sent to.  The default is blank.  
	If the server is using the "connected" ditribution type, a blank address will 
	make the client's share of generated coins be kept by the server.  If the 
	server is using the "contributed" distribution type, a blank address will
	mean the client contributes as normal, but the contribution is ignored when
	determining how to distribute any coins.

-threads=x
	Start this number of miner threads.  The default value is the number of cores
	on your processor if using the CPU miner, or 1 if using a GPU miner.


*********************
* CUDA MINER
*********************
CUDA miner arguments

-gpu=X
	Turns on GPU processing on specific GPU device.  Indexes start at 0.  If you 
	just use -gpu without =X it will pick the device with the max GFlops.

-aggression=X
	Specifies how many hashes (2^X) per kernel thread will be calculated.  
	Default is 6.  It starts at 1 and goes to 32, with each successive number 
	meaning double the number of hashes.  Sane values are 1 to 12 or maybe 14 if 
	you have some super card.

-gpugrid=X
	Specifies what the grid size of the kernel should be.  Useful for fine tuning 
	hash rate.

-gputhreads=X
	Specifies how many threads per kernel invocation should run.  Useful for fine 
	tuning hash rate.

-port=X
	Specifies the port that bitcoin will listen on.  (When run in GUI or daemon)

-rpcport=X
	Specifies the port that the rpc server will listen on.  (When run in GUI or daemon)



*********************
* OPENCL MINER
*********************
Make sure bitcoinmineropencl.cl is in the bitcoin working directory.

OpenCL miner arguments

-platform=X
	Use specific OpenCL platform at index X.  Indexes start at 0.  Default is 0.
	
-gpu=X
	Turns on GPU processing on specific GPU device on specified platform.
	Indexes start at 0.  If you just use -gpu without =X it will pick the first 
	device found.

-aggression=X
	Specifies how many hashes (2^X) per kernel thread will be calculated.  
	Default is 6.  It starts at 1 and goes to 32, with each successive number 
	meaning double the number of hashes.  Sane values are 1 to 12 or maybe 14 if 
	you have some super card.

-gpugrid=X
	Specifies what the grid size of the kernel should be.  Useful for fine tuning 
	hash rate.

-gputhreads=X
	Specifies how many threads per kernel invocation should run.  Useful for fine 
	tuning hash rate.

-port=X
	Specifies the port that bitcoin will listen on.  (When run in GUI or daemon)

-rpcport=X
	Specifies the port that the rpc server will listen on.  (When run in GUI or daemon)


debian/changelog
bitcoinr-miner (0.0.20101211-1) unstable; urgency=low

  * Initial Release.

 -- Mario Oyorzabal Salgado <tuxsoul@tuxsoul.com>  Fri,  3 Dec 2010 23:59:59 -0600


src/remote/remoteminer.h
/**
    Copyright (C) 2010  puddinpop

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License along
    with this program; if not, write to the Free Software Foundation, Inc.,
    51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
**/

#ifndef _bitcoin_remote_miner_
#define _bitcoin_remote_miner_

#include "../headers.h"
#include "remoteminermessage.h"
#include "../cryptopp/sha.h"
#include <vector>
#include <map>

extern const int BITCOINMINERREMOTE_THREADINDEX;
extern const int BITCOINMINERREMOTE_HASHESPERMETA;
#define BITCOINMINERREMOTE_SERVERVERSIONSTR "1.2.0"

void ThreadBitcoinMinerRemote(void* parg);
void BitcoinMinerRemote();

extern CCriticalSection cs_mapTransactions;
unsigned int GetNextWorkRequired(const CBlockIndex* pindexLast);
int FormatHashBlocks(void* pbuffer, unsigned int len);
void SHA256Transform(void* pstate, void* pinput, const void* pinit);
bool ProcessBlock(CNode* pfrom, CBlock* pblock);

class RemoteClientConnection
{
public:
	RemoteClientConnection(const SOCKET sock, sockaddr_storage &addr, const int addrlen);
	~RemoteClientConnection();

	const SOCKET GetSocket() const		{ return m_socket; }

	const bool IsConnected() const		{ return m_socket!=INVALID_SOCKET; }
	const bool Disconnect();

	void SendMessage(const RemoteMinerMessage &message);
	const bool MessageReady() const;
	const bool ProtocolError() const;
	const bool ReceiveMessage(RemoteMinerMessage &message);

	void SetRequestedRecipientAddress(const uint160 &recipient)		{ m_recipientaddress=recipient; }
	const bool GetRequestedRecipientAddress(uint160 &recipient)		{ recipient=m_recipientaddress; return m_recipientaddress!=0; }
	const int64 GetCalculatedKHashRateFromBestHash(const int sec=60, const int minsec=60) const;
	const int64 GetCalculatedKHashRateFromMetaHash(const int sec=60, const int minsec=60) const;
	const int64 GetReportedKHashRate() const;
	const bool GotClientHello() const								{ return m_gotclienthello; }
	void SetGotClientHello(bool got)								{ m_gotclienthello=got; }

	const time_t GetLastVerifiedMetaHash() const					{ return m_lastverifiedmetahash; }
	void SetLastVerifiedMetaHash(const time_t t)					{ m_lastverifiedmetahash=t; }
	const int64 GetVerifiedMetaHashCount() const					{ return m_verifiedmetahashcount; }

	const std::vector<char>::size_type ReceiveBufferSize() const	{ return m_receivebuffer.size(); }
	const std::vector<char>::size_type SendBufferSize() const		{ return m_sendbuffer.size(); }

	const bool SocketReceive();
	const bool SocketSend();

	const std::string GetAddress(const bool withport=true) const;

	const uint160 GetRecipientAddress() const						{ return m_recipientaddress; }

	int64 &NextBlockID()											{ return m_nextblockid; }

	void ClearOldSentWork(const int sec);

	struct metahash
	{
		metahash():m_verified(false),m_senttime(time(0))	{ }
		std::vector<unsigned char> m_metahash;
		unsigned int m_startnonce;
		bool m_verified;
		time_t m_senttime;
		uint256 m_besthash;
		unsigned int m_besthashnonce;
	};

	struct sentwork
	{
		sentwork():m_pblock(0)			{ }
		
		int64 m_blockid;
		time_t m_senttime;
		std::vector<unsigned char> m_block;
		std::vector<unsigned char> m_midstate;
		uint256 m_target;
		CKey m_key;
		std::vector<metahash> m_metahashes;
		CBlock *m_pblock;
		CBlockIndex *m_indexprev;

		const bool CheckNonceOverlap(const unsigned int nonce, const unsigned int hashespermetahash)
		{
			for(std::vector<metahash>::const_iterator i=m_metahashes.begin(); i!=m_metahashes.end(); i++)
			{
				if((*i).m_startnonce<=nonce && (*i).m_startnonce+hashespermetahash>nonce)
				{
					return true;
				}
			}
			return false;
		}
	};

	const std::vector<sentwork> &GetSentWork() const		{ return m_sentwork; }
	std::vector<sentwork> &GetSentWork()					{ return m_sentwork; }
	const bool GetSentWorkByBlock(const std::vector<unsigned char> &block, sentwork **work);
	const bool GetSentWorkByID(const int64 id, sentwork **work);
	const bool GetNewestSentWorkWithMetaHash(sentwork &work) const;
	void SetWorkVerified(sentwork &work);

private:
	SOCKET m_socket;
	struct sockaddr_storage m_addr;
	int m_addrlen;

	std::vector<char> m_tempbuffer;
	std::vector<char> m_receivebuffer;
	std::vector<char> m_sendbuffer;

	std::vector<sentwork> m_sentwork;

	time_t m_connecttime;
	time_t m_lastactive;
	time_t m_lastverifiedmetahash;
	bool m_gotclienthello;
	uint160 m_recipientaddress;

	int64 m_nextblockid;
	int64 m_verifiedmetahashcount;

};

class BitcoinMinerRemoteServer
{
public:
	BitcoinMinerRemoteServer();
	~BitcoinMinerRemoteServer();

	const bool StartListen(const std::string &bindaddr="127.0.0.1", const std::string &bindport="8335");
	const bool Step();

	std::vector<RemoteClientConnection *> &Clients()		{ return m_clients; }

	void SendServerHello(RemoteClientConnection *client, const int metahashrate);
	void SendWork(RemoteClientConnection *client);
	void SendServerStatus();
	void SendWorkToAllClients();

	static const bool EncodeBase64(const std::vector<unsigned char> &data, std::string &encoded);
	static const bool DecodeBase64(const std::string &encoded, std::vector<unsigned char> &decoded);

	const int64 GetAllClientsCalculatedKHashFromMeta() const;
	const int64 GetAllClientsCalculatedKHashFromBest() const;
	
	RemoteClientConnection *GetOldestNonVerifiedMetaHashClient();

	int64 &GeneratedCount()													{ return m_generatedcount; }

	void LoadContributedHashes();
	void SaveContributedHashes();

	void AddContributedHashes(const uint160 address, const int64 hashes)	{ m_currenthashescontributed[address]+=hashes; }
	void ClearCurrentHashesContributed()									{ m_previoushashescontributed=m_currenthashescontributed; m_currenthashescontributed.clear(); }

private:
	void BlockToJson(const CBlock *block, json_spirit::Object &obj);
	void ReadBanned(const std::string &filename);

	void AddDistributionFromConnected(CBlock *pblock, CBlockIndex *pindexPrev, int64 nFees);
	void AddDistributionFromContributed(CBlock *pblock, CBlockIndex *pindexPrev, int64 nFees);

	CBigNum m_bnExtraNonce;

#ifdef _WIN32
	static bool m_wsastartup;
#endif
	std::string m_distributiontype;
	std::vector<SOCKET> m_listensockets;
	std::vector<RemoteClientConnection *> m_clients;
	std::map<uint160,uint256> m_previoushashescontributed;	// number of hashes each address contributed to the previous block solve
	std::map<uint160,uint256> m_currenthashescontributed;		// number of hashes each address contributed since last block solve
	std::set<std::string> m_banned;
	time_t m_startuptime;
	int64 m_generatedcount;

};

#endif	// _bitcoin_remote_miner_


src/remote/remoteminerclient.cpp
/**
    Copyright (C) 2010  puddinpop

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License along
    with this program; if not, write to the Free Software Foundation, Inc.,
    51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
**/

#include "remoteminerclient.h"
#include "remoteminermessage.h"
#include "base64.h"

#include <iostream>
#include <string>

#ifdef _WIN32
	#include <winsock2.h>
	#include <ws2tcpip.h>
#else
	#include <arpa/inet.h>
	#include <netinet/in.h>
	#include <netdb.h>
#endif

#ifdef _WIN32
bool RemoteMinerClient::m_wsastartup=false;
#endif

RemoteMinerClient::RemoteMinerClient():m_socket(INVALID_SOCKET),m_tempbuffer(4096,0),m_gotserverhello(false)
{
#ifdef _WIN32
	if(m_wsastartup==false)
	{
		WSAData wsadata;
		WSAStartup(MAKEWORD(2,2),&wsadata);
		m_wsastartup=true;
	}
#endif
	m_midbuffptr=alignup<16>(m_currentmidbuff);
	m_blockbuffptr=alignup<16>(m_currentblockbuff);
	m_nonce=(unsigned int *)(m_blockbuffptr+12);
	m_nextblockid=0;
	m_currentblockid=0;
}

RemoteMinerClient::~RemoteMinerClient()
{
	Disconnect();
}

const bool RemoteMinerClient::Connect(const std::string &server, const std::string &port)
{
	m_sendbuffer.clear();
	m_receivebuffer.clear();

	if(IsConnected()==true)
	{
		Disconnect();
	}

	int rval=-1;
	std::ostringstream portstring;
	addrinfo hint,*result,*current;

	result=current=0;
	std::memset(&hint,0,sizeof(hint));
	hint.ai_socktype=SOCK_STREAM;

	rval=getaddrinfo(server.c_str(),port.c_str(),&hint,&result);

	if(result)
	{
		for(current=result; current!=0 && m_socket==INVALID_SOCKET; current=current->ai_next)
		{

			m_socket=socket(current->ai_family,current->ai_socktype,current->ai_protocol);

			if(m_socket!=INVALID_SOCKET)
			{
				rval=connect(m_socket,current->ai_addr,current->ai_addrlen);
				if(rval==-1)
				{
					Disconnect();
				}
			}

		}

		freeaddrinfo(result);
	}

	if(rval==0)
	{
		return true;
	}
	else
	{
		return false;
	}
}

const bool RemoteMinerClient::DecodeBase64(const std::string &encoded, std::vector<unsigned char> &decoded) const
{
	if(encoded.size()>0)
	{
		int dlen=((encoded.size()*3)/4)+4;
		decoded.resize(dlen,0);
		std::vector<unsigned char> src(encoded.begin(),encoded.end());
		if(base64_decode(&decoded[0],&dlen,&src[0],src.size())==0)
		{
			decoded.resize(dlen);
			return true;
		}
		else
		{
			return false;
		}
	}
	else
	{
		decoded.resize(0);
		return true;
	}
}

const bool RemoteMinerClient::Disconnect()
{
	m_sendbuffer.clear();
	m_receivebuffer.clear();
	if(IsConnected())
	{
		myclosesocket(m_socket);
		m_socket=INVALID_SOCKET;
	}
	return true;
}

const bool RemoteMinerClient::EncodeBase64(const std::vector<unsigned char> &data, std::string &encoded) const
{
	if(data.size()>0)
	{
		int dstlen=((data.size()*4)/3)+4;
		std::vector<unsigned char> dst(dstlen,0);
		if(base64_encode(&dst[0],&dstlen,&data[0],data.size())==0)
		{
			dst.resize(dstlen);
			encoded.assign(dst.begin(),dst.end());
			return true;
		}
		else
		{
			return false;
		}
	}
	else
	{
		encoded=std::string("");
		return true;
	}
}

const bool RemoteMinerClient::FindGenerationAddressInBlock(const uint160 address, json_spirit::Object &obj, double &amount) const
{

	json_spirit::Value tx=json_spirit::find_value(obj,"tx");
	if(tx.type()==json_spirit::array_type)
	{
		json_spirit::Array txarray=tx.get_array();
		for(json_spirit::Array::iterator i=txarray.begin(); i!=txarray.end(); i++)
		{
			bool isgenerationtx=false;
			json_spirit::Value in=json_spirit::find_value((*i).get_obj(),"in");
			if(in.type()==json_spirit::array_type)
			{
				json_spirit::Array inarray=in.get_array();
				for(json_spirit::Array::iterator j=inarray.begin(); j!=inarray.end(); j++)
				{
					json_spirit::Value prev=json_spirit::find_value((*j).get_obj(),"prev_out");
					if(prev.type()==json_spirit::obj_type)
					{
						json_spirit::Value hash=json_spirit::find_value(prev.get_obj(),"hash");
						if(hash.type()==json_spirit::str_type)
						{
							uint256 h(hash.get_str());
							if(h==0)
							{
								isgenerationtx=true;
							}
						}
					}
				}
			}

			if(isgenerationtx==true)
			{
				json_spirit::Value out=json_spirit::find_value((*i).get_obj(),"out");
				if(out.type()==json_spirit::array_type)
				{
					json_spirit::Array outarray=out.get_array();
					for(json_spirit::Array::iterator j=outarray.begin(); j!=outarray.end(); j++)
					{
						json_spirit::Value script=json_spirit::find_value((*j).get_obj(),"scriptPubKey");
						if(script.type()==json_spirit::str_type)
						{
							std::string scr=script.get_str();
							if(scr.find(ReverseAddressHex(address))!=std::string::npos)
							{
								json_spirit::Value val=json_spirit::find_value((*j).get_obj(),"value");
								if(val.type()==json_spirit::real_type)
								{
									amount=val.get_real();
									return true;
								}
							}
						}
					}
				}
			}

		}
	}

	return false;
}

void RemoteMinerClient::HandleMessage(const RemoteMinerMessage &message)
{
	json_spirit::Value tval=json_spirit::find_value(message.GetValue().get_obj(),"type");
	if(tval.type()==json_spirit::int_type)
	{
		std::cout << "Got message " << tval.get_int() << " from server." << std::endl;
		if(tval.get_int()==RemoteMinerMessage::MESSAGE_TYPE_SERVERHELLO)
		{
			m_gotserverhello=true;
			tval=json_spirit::find_value(message.GetValue().get_obj(),"metahashrate");
			if(tval.type()==json_spirit::int_type)
			{
				m_metahashstartblock.clear();
				m_metahashstartblock.insert(m_metahashstartblock.end(),128,0);
				m_metahash.clear();
				m_metahash.resize(tval.get_int());
				m_metahashpos=0;
			}
			tval=json_spirit::find_value(message.GetValue().get_obj(),"serverversion");
			if(tval.type()==json_spirit::str_type)
			{
				std::cout << "Server version " << tval.get_str() << std::endl;
			}
		}
		else if(tval.get_int()==RemoteMinerMessage::MESSAGE_TYPE_SERVERSENDWORK)
		{
			tval=json_spirit::find_value(message.GetValue().get_obj(),"blockid");
			if(tval.type()==json_spirit::int_type)
			{
				m_nextblockid=tval.get_int();
			}
			tval=json_spirit::find_value(message.GetValue().get_obj(),"block");
			if(tval.type()==json_spirit::str_type)
			{
				DecodeBase64(tval.get_str(),m_nextblock);
			}
			tval=json_spirit::find_value(message.GetValue().get_obj(),"target");
			if(tval.type()==json_spirit::str_type)
			{
				m_nexttarget.SetHex(tval.get_str());
			}
			tval=json_spirit::find_value(message.GetValue().get_obj(),"midstate");
			if(tval.type()==json_spirit::str_type)
			{
				DecodeBase64(tval.get_str(),m_nextmidstate);
			}

			tval=json_spirit::find_value(message.GetValue().get_obj(),"fullblock");
			if(m_address160!=0 && tval.type()==json_spirit::obj_type)
			{
				double amount=0;
				if(FindGenerationAddressInBlock(m_address160,tval.get_obj(),amount))
				{
					std::cout << "Address " << Hash160ToAddress(m_address160) <<  " will receive " << amount << " BTC if this block is solved" << std::endl;
				}
				else
				{
					std::cout << "Address " << Hash160ToAddress(m_address160) << " not found in block being solved" << std::endl;
				}
			}

			if(m_havework==false)
			{
				m_currenttarget=m_nexttarget;
				m_currentblockid=m_nextblockid;
				::memcpy(m_midbuffptr,&m_nextmidstate[0],32);
				::memcpy(m_blockbuffptr,&m_nextblock[0],64);
				m_metahashstartblock=m_nextblock;
				m_metahashpos=0;
				m_metahashstartnonce=0;
				(*m_nonce)=0;
			}
			m_havework=true;
		}
		else if(tval.get_int()==RemoteMinerMessage::MESSAGE_TYPE_SERVERSTATUS)
		{
			int64 clients=0;
			int64 khashmeta=0;
			int64 khashbest=0;
			int64 clientkhashmeta=0;
			time_t startuptime=0;
			struct tm startuptimetm;
			std::string startuptimestr("");
			int64 blocksgenerated=0;

			tval=json_spirit::find_value(message.GetValue().get_obj(),"clients");
			if(tval.type()==json_spirit::int_type)
			{
				clients=tval.get_int();
			}
			tval=json_spirit::find_value(message.GetValue().get_obj(),"khashmeta");
			if(tval.type()==json_spirit::int_type)
			{
				khashmeta=tval.get_int();
			}
			tval=json_spirit::find_value(message.GetValue().get_obj(),"khashbest");
			if(tval.type()==json_spirit::int_type)
			{
				khashbest=tval.get_int();
			}
			tval=json_spirit::find_value(message.GetValue().get_obj(),"yourkhashmeta");
			if(tval.type()==json_spirit::int_type)
			{
				clientkhashmeta=tval.get_int();
			}
			tval=json_spirit::find_value(message.GetValue().get_obj(),"sessionstartuptime");
			if(tval.type()==json_spirit::int_type)
			{
				startuptime=tval.get_int();
				startuptimetm=*gmtime(&startuptime);
				std::vector<char> buff(128,0);
				int rval=strftime(&buff[0],buff.size()-1,"%Y-%m-%d %H:%M:%S",&startuptimetm);
				buff.resize(rval);
				startuptimestr=std::string(buff.begin(),buff.end());
			}
			tval=json_spirit::find_value(message.GetValue().get_obj(),"sessionblocksgenerated");
			if(tval.type()==json_spirit::int_type)
			{
				blocksgenerated=tval.get_int();
			}
			
			//std::cout << "Server Status : " << clients << " clients, " << khashmeta << " khash/s m " << khashbest << " khash/s b  " << std::endl;
			std::cout << "Server Status : " << clients << " clients, " << khashmeta << " khash/s" << std::endl;
			std::cout << blocksgenerated << " blocks generated since " << startuptimestr << " UTC" << std::endl;
			std::cout << "Server reports my khash/s as " << clientkhashmeta << std::endl;
		}
	}
	else
	{
		std::cout << "Server sent invalid message.  Disconnecting." << std::endl;
	}
}

const bool RemoteMinerClient::MessageReady() const
{
	return RemoteMinerMessage::MessageReady(m_receivebuffer);
}

const bool RemoteMinerClient::ProtocolError() const
{
	return RemoteMinerMessage::ProtocolError(m_receivebuffer);
}

const bool RemoteMinerClient::ReceiveMessage(RemoteMinerMessage &message)
{
	return RemoteMinerMessage::ReceiveMessage(m_receivebuffer,message);
}

const std::string RemoteMinerClient::ReverseAddressHex(const uint160 address) const
{
	std::string rval("");
	std::string addresshex=address.GetHex();
	for(std::string::size_type i=0; i<addresshex.size(); i+=2)
	{
		if(i+1<addresshex.size())
		{
			rval=addresshex.substr(i,2)+rval;
		}
		else
		{
			rval=addresshex.substr(i,1)+rval;
		}
	}
	return rval;
}

void RemoteMinerClient::Run(const std::string &server, const std::string &port, const std::string &password, const std::string &address)
{
	static const unsigned int SHA256InitState[8] ={0x6a09e667, 0xbb67ae85, 0x3c6ef372, 0xa54ff53a, 0x510e527f, 0x9b05688c, 0x1f83d9ab, 0x5be0cd19};
	uint256 tempbuff[4];
	uint256 &temphash=*alignup<16>(tempbuff);
	uint256 hashbuff[4];
	uint256 &hash=*alignup<16>(hashbuff);
	uint256 lasttarget=0;

	uint256 besthash=~(uint256(0));
	unsigned int besthashnonce=0;

	SetThreadPriority(THREAD_PRIORITY_LOWEST);

	FormatHashBlocks(&temphash,sizeof(temphash));
	for(int i=0; i<64/4; i++)
	{
		((unsigned int*)&temphash)[i] = CryptoPP::ByteReverse(((unsigned int*)&temphash)[i]);
	}

	while(true)
	{
		if(IsConnected()==false)
		{
			std::cout << "Attempting to connect to " << server << ":" << port << std::endl;
			Sleep(1000);
			if(Connect(server,port))
			{
				m_gotserverhello=false;
				m_havework=false;
				m_metahashpos=0;
				lasttarget=0;
				(*m_nonce)=0;
				m_metahashstartnonce=0;
				besthash=~(uint256(0));
				besthashnonce=0;
				std::cout << "Connected to " << server << ":" << port << std::endl;
				SendClientHello(password,address);
			}
		}
		else
		{
			if(m_gotserverhello==false || m_metahashpos%50000==0)
			{
				Update();
				while(MessageReady() && !ProtocolError())
				{
					RemoteMinerMessage message;
					if(ReceiveMessage(message))
					{
						if(message.GetValue().type()==json_spirit::obj_type)
						{
							HandleMessage(message);
						}
						else
						{
							std::cout << "Unexpected json type sent by server.  Disconnecting." << std::endl;
							Disconnect();
						}
					}
				}
				if(ProtocolError())
				{
					std::cout << "Protocol error.  Disconnecting." << std::endl;
					Disconnect();
				}
			}

			if(m_havework)
			{
				// do 10000 hashes at a time
				for(unsigned int i=0; i<10000 && m_metahashpos<m_metahash.size(); i++)
				{
					SHA256Transform(&temphash,m_blockbuffptr,m_midbuffptr);
					SHA256Transform(&hash,&temphash,SHA256InitState);

					m_metahash[m_metahashpos++]=((unsigned char *)&hash)[0];

					if((((unsigned short*)&hash)[14]==0) && (((unsigned short*)&hash)[15]==0))
					{
						for (int i = 0; i < sizeof(hash)/4; i++)
						{
							((unsigned int*)&hash)[i] = CryptoPP::ByteReverse(((unsigned int*)&hash)[i]);
						}
						
						if(hash<=m_currenttarget)
						{
							std::cout << "Found block!" << std::endl;

							// send hash found to server
							SendFoundHash(m_currentblockid,m_metahashstartblock,(*m_nonce));
						}

						if(hash<besthash)
						{
							besthash=hash;
							besthashnonce=(*m_nonce);
						}
					}
					// hash isn't bytereversed yet, but besthash already is
					else if(CryptoPP::ByteReverse(((unsigned int*)&hash)[7])<=((unsigned int *)&besthash)[7])
					{
						for (int i = 0; i < sizeof(hash)/4; i++)
						{
							((unsigned int*)&hash)[i] = CryptoPP::ByteReverse(((unsigned int*)&hash)[i]);
						}
						if(hash<besthash)
						{
							besthash=hash;
							besthashnonce=(*m_nonce);
						}
					}

					(*m_nonce)++;
				}

				if(m_metahashpos>=m_metahash.size())
				{

					std::vector<unsigned char> digest(SHA256_DIGEST_LENGTH,0);
					SHA256(&m_metahash[0],m_metahash.size(),&digest[0]);

					// send the metahash to the server and request a new block if the nonce is approaching the limit
					SendMetaHash(m_currentblockid,m_metahashstartblock,m_metahashstartnonce,digest,besthash,besthashnonce);
					if((*m_nonce)>4000000000)
					{
						std::cout << "Requesting a new block" << std::endl;
						SendWorkRequest();
					}

					// reset the nonce to 0 if this is a new block
					if(m_nextblock!=m_metahashstartblock)
					{
						(*m_nonce)=0;
					}
					// set meta nonce first because memcpy will overwrite it
					m_metahashstartnonce=(*m_nonce);
					m_currentblockid=m_nextblockid;
					m_currenttarget=m_nexttarget;
					::memcpy(m_midbuffptr,&m_nextmidstate[0],32);
					::memcpy(m_blockbuffptr,&m_nextblock[0],64);
					m_metahashstartblock=m_nextblock;
					m_metahashpos=0;
					// set nonce again because it was overwritten by memcpy
					(*m_nonce)=m_metahashstartnonce;

					if(lasttarget!=m_currenttarget)
					{
						std::cout << "Target Hash : " << m_currenttarget.ToString() << std::endl;
						lasttarget=m_currenttarget;
					}

					std::cout << "Best : " << besthash.ToString() << std::endl;
					besthash=~(uint256(0));
					besthashnonce=0;
				}

			}

		}
	}

}

void RemoteMinerClient::SendClientHello(const std::string &password, const std::string &address)
{
	json_spirit::Object obj;
	
	obj.push_back(json_spirit::Pair("type",RemoteMinerMessage::MESSAGE_TYPE_CLIENTHELLO));
	obj.push_back(json_spirit::Pair("password",password));
	if(address!="")
	{
		uint160 h160;
		if(AddressToHash160(address.c_str(),h160))
		{
			m_address160=h160;
			obj.push_back(json_spirit::Pair("address",h160.GetHex()));
		}
	}
	
	SendMessage(RemoteMinerMessage(obj));
}

void RemoteMinerClient::SendFoundHash(const int64 blockid, const std::vector<unsigned char> &block, const unsigned int nonce)
{
	json_spirit::Object obj;
	std::string blockstr("");
	
	EncodeBase64(block,blockstr);
	
	obj.push_back(json_spirit::Pair("type",static_cast<int>(RemoteMinerMessage::MESSAGE_TYPE_CLIENTFOUNDHASH)));
	if(blockid!=0)
	{
		obj.push_back(json_spirit::Pair("blockid",static_cast<int64>(blockid)));
	}
	else
	{
		obj.push_back(json_spirit::Pair("block",blockstr));
	}
	obj.push_back(json_spirit::Pair("nonce",static_cast<int64>(nonce)));
	
	SendMessage(RemoteMinerMessage(obj));
}

void RemoteMinerClient::SendMessage(const RemoteMinerMessage &message)
{
	message.PushWireData(m_sendbuffer);
}

void RemoteMinerClient::SendMetaHash(const int64 blockid, const std::vector<unsigned char> &block, const unsigned int startnonce, const std::vector<unsigned char> &digest, const uint256 &besthash, const unsigned int besthashnonce)
{
	std::string blockstr("");
	std::string digeststr("");

	EncodeBase64(block,blockstr);
	EncodeBase64(digest,digeststr);
	json_spirit::Object obj;
	obj.push_back(json_spirit::Pair("type",static_cast<int>(RemoteMinerMessage::MESSAGE_TYPE_CLIENTMETAHASH)));
	if(blockid!=0)
	{
		obj.push_back(json_spirit::Pair("blockid",static_cast<int64>(blockid)));
	}
	else
	{
		obj.push_back(json_spirit::Pair("block",blockstr));
	}
	obj.push_back(json_spirit::Pair("nonce",static_cast<int64>(startnonce)));
	obj.push_back(json_spirit::Pair("digest",digeststr));
	obj.push_back(json_spirit::Pair("besthash",besthash.ToString()));
	obj.push_back(json_spirit::Pair("besthashnonce",static_cast<int64>(besthashnonce)));

	SendMessage(RemoteMinerMessage(obj));
}

void RemoteMinerClient::SendWorkRequest()
{
	json_spirit::Object obj;
	obj.push_back(json_spirit::Pair("type",RemoteMinerMessage::MESSAGE_TYPE_CLIENTGETWORK));

	SendMessage(RemoteMinerMessage(obj));
}

void RemoteMinerClient::SocketReceive()
{
	if(IsConnected())
	{
		int len=::recv(m_socket,&m_tempbuffer[0],m_tempbuffer.size(),0);
		if(len>0)
		{
			m_receivebuffer.insert(m_receivebuffer.end(),m_tempbuffer.begin(),m_tempbuffer.begin()+len);
		}
		else
		{
			Disconnect();
		}
	}
}

void RemoteMinerClient::SocketSend()
{
	if(IsConnected() && m_sendbuffer.size()>0)
	{
		int len=::send(m_socket,&m_sendbuffer[0],m_sendbuffer.size(),0);
		if(len>0)
		{
			m_sendbuffer.erase(m_sendbuffer.begin(),m_sendbuffer.begin()+len);
		}
		else
		{
			Disconnect();
		}
	}
}

const bool RemoteMinerClient::Update()
{
	if(IsConnected())
	{
		m_timeval.tv_sec=0;
		m_timeval.tv_usec=0;

		FD_ZERO(&m_readfs);
		FD_ZERO(&m_writefs);

		FD_SET(m_socket,&m_readfs);

		if(m_sendbuffer.size()>0)
		{
			FD_SET(m_socket,&m_writefs);
		}

		select(m_socket+1,&m_readfs,&m_writefs,0,&m_timeval);

		if(IsConnected() && FD_ISSET(m_socket,&m_readfs))
		{
			SocketReceive();
		}
		if(IsConnected() && FD_ISSET(m_socket,&m_writefs))
		{
			SocketSend();
		}

	}

	if(IsConnected())
	{
		return true;
	}
	else
	{
		return false;
	}

}
